src/main/webapp/signup.js

```diff
@@ -0,0 +1,162 @@
+} from './js/fetch_handler.js';
+
+const AFTER_SIGNUP_REDIRECT = '/dashboard.html';
+
+/**
+ * Initializes the sign up form.
+ */
+function initSignupForm() {
+  const signupForm = document.getElementById('signup-form');
+  signupForm.addEventListener('submit', submitSignup);
+
+  signupForm.elements['github-link-button'].addEventListener(
+      'click', handleGitHubLink);
+  gitHubAuthorizer
+      .getFirebase().auth().onAuthStateChanged(handleAuthStateChanged);
+  // Handle auth state changed on page load in case of existing authentication.
+  handleAuthStateChanged(gitHubAuthorizer.getUser());
+}
+
+/**
+ * Handles the click event for the GitHub linking button.
+ * @param {Event} e The button click event.
+ */
+function handleGitHubLink(e) {
+  e.preventDefault();
+
+  gitHubAuthorizer.toggleSignIn().catch((error) => {
+    console.error(error);
+    alert('Encountered an error with GitHub authentication.'
+        + ' Please try again later.');
+  });
+}
+
+/**
+ * Updates the signup UI when the state of user GitHub authentication changes.
+ * @param {Firebase.User} user
+ */
+function handleAuthStateChanged(user) {
+  const signupForm = document.getElementById('signup-form');
+  const submitButton = signupForm.elements['submit-button'];
+  const gitHubLinkButton = signupForm.elements['github-link-button'];
+  if (user) {
+    submitButton.disabled = false;
+    gitHubLinkButton.innerText = 'Unlink Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-success');
+    gitHubLinkButton.classList.remove('btn-purple');
+  } else {
+    submitButton.disabled = true;
+    gitHubLinkButton.innerText = 'Link Your GitHub Account';
+    gitHubLinkButton.classList.add('btn-purple');
+    gitHubLinkButton.classList.remove('btn-success');
+  }
+}
+
+/**
+ * Validates and submits a user signup request.
+ * @param {Event} e The signup form submit event.
+ */
+function submitSignup(e) {
+  e.preventDefault();
+
+  const signupForm = document.getElementById('signup-form');
+  const gitHubLinkButton = signupForm.elements['github-link-button'];
+  const submitButton = signupForm.elements['submit-button'];
+  gitHubLinkButton.disabled = true;
+  submitButton.disabled = true;
+
+  const signupBody = createSignupBody();
+  console.log('Body of signup request:');
+  console.log(signupBody);
+  if (signupBody) {
+    const signupRequest = createSignupRequest(signupBody);
+
+    signupRequest.then((response) => {
+      // If signup is successful, redirect.
+      gitHubLinkButton.disabled = false;
+      submitButton.disabled = false;
+      window.location.href = AFTER_SIGNUP_REDIRECT;
+    }).catch((errorResponse) => {
+      console.error(
+        `Error ${errorResponse.statusCode}: ${errorResponse.error}`);
+      alert(errorResponse.userMessage);
+      gitHubLinkButton.disabled = false;
+      submitButton.disabled = false;
+    });
+  } else {
+    gitHubLinkButton.disabled = false;
+    submitButton.disabled = false;
+  }
```

Comment:
```suggestion
  if (signupBody) {
    const signupRequest = createSignupRequest(signupBody);

    signupRequest.then((response) => {
      // If signup is successful, redirect.
      window.location.href = AFTER_SIGNUP_REDIRECT;
    }).catch((errorResponse) => {
      console.error(
        `Error ${errorResponse.statusCode}: ${errorResponse.error}`);
      alert(errorResponse.userMessage);
    });
  }
  gitHubLinkButton.disabled = false;
  submitButton.disabled = false;
```

Reply:
I agree that this would be an ideal change, although because `signupRequest.then((response) => [...])` is asynchronous, it would mean that the buttons would be re-enabled immediately following the start of the request, not the end of the request after the response has been received. Looking at this again now, I can consolidate the ones in the promise slightly and will do that momentarily.